### PR TITLE
ensure filehandles are closed

### PR DESCRIPTION
**Pull request by docwhat**
See https://www.joeshaw.org/dont-defer-close-on-writable-files/

### CODE REVIEW

**Comment by docwhat** [comment]
I can't reproduce this locally... I *think* it is a legit error:

```
=== FAIL: internal/filewatcher TestHasGoFiles/none (0.00s)
    --- FAIL: TestHasGoFiles/none (0.00s)
panic: sync C:\Users\circleci\AppData\Local\Temp\TestHasGoFiles-none-706743447: The handle is invalid. [recovered]
	panic: sync C:\Users\circleci\AppData\Local\Temp\TestHasGoFiles-none-706743447: The handle is invalid.

goroutine 16 [running]:
testing.tRunner.func1.1(0xc62da0, 0xc00005ee40)
	c:/go/src/testing/testing.go:1072 +0x310
testing.tRunner.func1(0xc00016c300)
	c:/go/src/testing/testing.go:1075 +0x43a
panic(0xc62da0, 0xc00005ee40)
	c:/go/src/runtime/panic.go:969 +0x1c7
gotest.tools/gotestsum/internal/filewatcher.hasGoFiles(0xc000012460, 0x42, 0xc00000a800)
	C:/Users/circleci/project/internal/filewatcher/watch.go:167 +0x493
gotest.tools/gotestsum/internal/filewatcher.TestHasGoFiles.func1(0xc00016c300)
	C:/Users/circleci/project/internal/filewatcher/watch_test.go:88 +0x115
testing.tRunner(0xc00016c300, 0xca6a70)
	c:/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	c:/go/src/testing/testing.go:1168 +0x2b3

=== FAIL: internal/filewatcher TestHasGoFiles (0.00s)
```

**Comment by docwhat** [comment]
I'd be okay with disabling at least this one due to [this reasoning](https://www.joeshaw.org/dont-defer-close-on-writable-files/).

```
internal/filewatcher/watch.go:160:16: Error return value of `fh.Close` is not checked (errcheck)
	defer fh.Close()
```

File: internal/filewatcher/watch.go [hunk 1]
```diff
@@ -157,19 +157,29 @@ func hasGoFiles(path string) bool {
 	if err != nil {
 		return false
 	}
+	defer fh.Close()
```
Comment (dnephin) [inline internal/filewatcher/watch.go:160]:
```suggestion
	defer fh.Close() // nolint: errcheck // fh is opened read-only
```

File: internal/filewatcher/watch.go [hunk 2]
```diff
@@ -157,19 +157,29 @@ func hasGoFiles(path string) bool {
...
 
 	for {
 		names, err := fh.Readdirnames(20)
 		switch {
 		case err == io.EOF:
+			if err := fh.Sync(); err != nil {
+				panic(err)
+			}
```
Comment (dnephin) [inline internal/filewatcher/watch.go:168]:
```suggestion
```

File: internal/filewatcher/watch.go [hunk 3]
```diff
@@ -157,19 +157,29 @@ func hasGoFiles(path string) bool {
...
 			return false
 		case err != nil:
 			log.Warnf("failed to read directory %v: %v", path, err)
+			if err := fh.Sync(); err != nil {
+				panic(err)
+			}
```
Comment (dnephin) [inline internal/filewatcher/watch.go:174]:
```suggestion
```

File: internal/filewatcher/watch.go [hunk 4]
```diff
@@ -157,19 +157,29 @@ func hasGoFiles(path string) bool {
...
 			return false
 		}
 
 		for _, name := range names {
 			if strings.HasSuffix(name, ".go") {
+				if err := fh.Sync(); err != nil {
+					panic(err)
+				}
```
Comment (dnephin) [inline internal/filewatcher/watch.go:182]:
```suggestion
```

**Review by dnephin — COMMENTED** [review]
Thank you for tracking down the problem and opening this PR! You are absolutely correct this missing `fh.Close()` is a problem. `golangci-lint` has linters for catching missing close on http response bodies, and sql rows, but apparently not for file handlers unfortunately.

That article is a great reminder about how to safely handle *writes*. Thankfully in this case we are only doing reads (https://golang.org/pkg/os/#Open, "Open opens the named file for reading"). I believe that explains the panic we see from windows. I guess on linux it is not an error to try and sync with a read-only handle, but on windows it is.

Since this is read-only we can remove the `fh.Sync()`, there isn't anything to sync.

I agree we can also ignore the lint error about the ignored return, we don't care about errors in this case because it is read-only.

I've added some suggested edits which should resolve the build errors. I'll apply those and merge once CI is happy.

**Comment by dnephin** [comment]
LGTM! CI is happy with those changes!

**Pull request merged**